builder/chroots.go

```diff
@@ -0,0 +1,637 @@
+func createVersionsFile(baseDir string, yumCmd []string) error {
+	// TODO: See if we query the list of packages some other way? Yum output is a bit
+	// unfriendly, see the workarounds below.
+	args := merge(yumCmd,
+		"--installroot="+filepath.Join(baseDir, "os-core"),
+		"list",
+	)
+
+	var outBuf bytes.Buffer
+	var errBuf bytes.Buffer
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Stdout = &outBuf
+	cmd.Stderr = &errBuf
+	err := cmd.Run()
+	if err != nil {
+		msg := fmt.Sprintf("couldn't list packages: %s\nCOMMAND LINE: %s", err, args)
+		if errBuf.Len() > 0 {
+			msg += "\nOUTPUT:\n%s" + errBuf.String()
+		}
+		return errors.New(msg)
+	}
+
+	type pkgEntry struct {
+		name, version string
+	}
+	var versions []*pkgEntry
+
+	scanner := bufio.NewScanner(&outBuf)
+	skippedPrefixes := []string{
+		// Default output from list command.
+		"Available",
+		"Installed",
+
+		// TODO: Review if those errors appear in stdout or stderr, if the former we can
+		// remove them. The rpm/yum cause the packages to be removed from the list.
+		"BDB2053", // Some Berkley DB error?
+		"rpm",
+		"yum",
+	}
+	for scanner.Scan() {
+		text := scanner.Text()
+
+		var skip bool
+		for _, p := range skippedPrefixes {
+			if strings.HasPrefix(text, p) {
+				skip = true
+				break
+			}
+		}
+		if skip {
+			continue
+		}
+
+		fields := strings.Fields(text)
+		if len(fields) != 3 {
+			// The output for yum list wraps at 80 when lacking information about the
+			// terminal, so we workaround by joining the next line and evaluating. See
+			// https://bugzilla.redhat.com/show_bug.cgi?id=584525 for the wrapping.
+			if scanner.Scan() {
+				text = text + scanner.Text()
+			} else {
+				return fmt.Errorf("couldn't parse line %q from yum list output", text)
+			}
+			fields = strings.Fields(text)
+			if len(fields) != 3 {
+				return fmt.Errorf("couldn't parse merged line %q from yum list output", text)
+			}
+		}
+
+		e := &pkgEntry{
+			name:    fields[0],
+			version: fields[1],
+		}
+		versions = append(versions, e)
+	}
+	err = scanner.Err()
+	if err != nil {
+		return err
+	}
+
+	sort.Slice(versions, func(i, j int) bool {
+		ii := versions[i]
+		jj := versions[j]
+		if ii.name == jj.name {
+			return ii.version < jj.version
+		}
+		return ii.name < jj.name
+	})
+
+	f, err := os.Create(filepath.Join(baseDir, "versions"))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = f.Close()
+	}()
+
+	w := bufio.NewWriter(f)
+	for _, e := range versions {
+		// TODO: change users of "versions" file to not rely on this exact formatting (version
+		// starting at column 51). E.g. this doesn't handle very well packages with large names.
+		fmt.Fprintf(w, "%-50s%s\n", e.name, e.version)
+	}
+	return w.Flush()
+}
+
+func fixOSRelease(filename, version string) error {
+	f, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = f.Close()
+	}()
+
+	// TODO: If this is a mix, NAME and ID should probably change too. Create a section in
```

Comment:
I think this is a good opportunity to update the os-release file to be more descriptive if it's on a mix, updating the name & adding a UpstreamVersion field perhaps.

Reply:
I agree, but think we can do that in a separate change. Created #113 and assigned to me.

Reply:
ok +1